File: app/src/main/java/xyz/untan/gyazotest/AuthActivity.java
package xyz.untan.gyazotest;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class AuthActivity extends AppCompatActivity {
    // SharedPreference util
    private AppStatus _appStatus;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // initialize member
        _appStatus = new AppStatus();

        Button loginView = (Button) findViewById(R.id.button_login);
        loginView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(GyazoApi.getAuthorizeIntent());
                finish(); // make sure don't get back to this activity
            }
        });

        if (_appStatus.accessToken != null && !_appStatus.accessToken.isEmpty()) {
            startActivity(new Intent(this, SettingsActivity.class));
            finish();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();

        Uri uri = getIntent().getData();
        if (uri != null) { // callback from gyazo
            onIntentCallback(uri);
        }
    }

    private void onIntentCallback(Uri uri) {
        // get authorization code
        String code = uri.getQueryParameter("code");
        if (code == null || code.isEmpty()) {
            showErrorToast();
            return;
        }

        GyazoApi.Api api = GyazoApi.getApi();
        api.token(code).enqueue(new Callback<GyazoApi.Api.TokenResponse>() {
            @Override
            public void onResponse(Call<GyazoApi.Api.TokenResponse> call,
                                   Response<GyazoApi.Api.TokenResponse> response) {
                if (!response.isSuccessful()) {
                    showErrorToast();
                    return;
                }

                _appStatus.accessToken = response.body().accessToken;
                _appStatus.save();

                startActivity(new Intent(AuthActivity.this, SettingsActivity.class));
                finish(); // make sure don't get back to this activity
            }

            @Override
            public void onFailure(Call<GyazoApi.Api.TokenResponse> call, Throwable t) {
                t.printStackTrace();
                showErrorToast();
            }
        });
    }

    private void showErrorToast() {
        // TODO custom style
        Toast.makeText(this, R.string.error_auth_failed, Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/java/xyz/untan/gyazauto/AuthActivity.kt
package xyz.untan.gyazauto

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.widget.Button
import android.widget.Toast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class AuthActivity : AppCompatActivity() {
    // SharedPreference util
    private var _appStatus: AppStatus? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // initialize member
        _appStatus = AppStatus()

        val loginView = findViewById(R.id.button_login) as Button
        loginView.setOnClickListener {
            startActivity(GyazoApi.authorizeIntent)
            finish() // make sure don't get back to this activity
        }

        if (!_appStatus!!.accessToken.isNullOrEmpty()) {
            startActivity(Intent(this, SettingsActivity::class.java))
            finish()
        }
    }

    override fun onStart() {
        super.onStart()

        val uri: Uri? = intent.data
        if (uri != null) { // callback from gyazo
            onIntentCallback(uri)
        }
    }

    private fun onIntentCallback(uri: Uri) {
        // get authorization code
        val code = uri.getQueryParameter("code")
        if (code.isEmpty()) {
            showErrorToast()
            return
        }

        val api = GyazoApi.api
        api.token(code).enqueue(object : Callback<GyazoApi.Api.TokenResponse> {
            override fun onResponse(call: Call<GyazoApi.Api.TokenResponse>,
                                    response: Response<GyazoApi.Api.TokenResponse>) {
                if (!response.isSuccessful) {
                    showErrorToast()
                    return
                }

                _appStatus!!.accessToken = response.body().accessToken
                _appStatus!!.save()

                startActivity(Intent(this@AuthActivity, SettingsActivity::class.java))
                finish() // make sure don't get back to this activity
            }

            override fun onFailure(call: Call<GyazoApi.Api.TokenResponse>, t: Throwable) {
                t.printStackTrace()
                showErrorToast()
            }
        })
    }

    private fun showErrorToast() {
        // TODO custom style
        Toast.makeText(this, R.string.error_auth_failed, Toast.LENGTH_LONG).show()
    }
}
